lib/promise/group.rb

```diff
@@ -1,59 +1,75 @@
 class Promise
   class Group
-    attr_accessor :source
     attr_reader :promise
 
-    def initialize(result_promise, inputs)
-      @promise = result_promise
-      @inputs = inputs
-      @remaining = count_promises
-      if @remaining.zero?
-        promise.fulfill(inputs)
-      else
-        promise.source = self
-        chain_inputs
-      end
+    def initialize(promise, values)
+      promise.source = self
+
+      @promise = promise
+      @values = values.dup
+      @total_resolved = 0
+
+      iterate
     end
 
     def wait
-      each_promise do |input_promise|
-        input_promise.wait if input_promise.pending?
+      return if resolved?
+
+      @values.each do |obj|
+        obj.wait if obj.is_a?(Promise) && obj.pending?
       end
     end
 
+    protected
+
+    def promise_fulfilled(value, index)
+      @total_resolved += 1
+      @values[index] = value
+
+      fulfill if resolved?
+    end
+
+    def promise_rejected(reason, _)
+      reject(reason)
+    end
+
     private
 
-    def chain_inputs
-      on_fulfill = method(:on_fulfill)
-      on_reject = promise.public_method(:reject)
-      each_promise do |input_promise|
-        input_promise.then(on_fulfill, on_reject)
-      end
+    def resolved?
+      @values && @values.length == @total_resolved
     end
 
-    def on_fulfill(_result)
-      @remaining -= 1
-      if @remaining.zero?
-        result = @inputs.map { |obj| promise?(obj) ? obj.value : obj }
-        promise.fulfill(result)
+    def iterate
+      @values.each_with_index do |maybe_promise, index|
```

Comment:
I'm not sure why, since it is easy to write an implementation in ruby that avoids object allocations, but here is where it happens in the source code: https://github.com/ruby/ruby/blob/v2_4_1/enum.c#L2214

This seems to be the case for the Enumerable methods and not for the Array methods.